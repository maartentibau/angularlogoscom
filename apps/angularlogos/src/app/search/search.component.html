<mat-form-field subscriptSizing="dynamic">
  <mat-label>Search</mat-label>
  <input matInput subscriptSizing="0" [formControl]="searchControl" />
  <button *ngIf="searchControl.value" matSuffix mat-icon-button aria-label="Clear" (click)="clear()">
    <mat-icon>close</mat-icon>
  </button>
</mat-form-field>
