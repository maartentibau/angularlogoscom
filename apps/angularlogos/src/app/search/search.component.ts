import { NgIf } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, OnChanges, Output } from '@angular/core';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { Observable } from 'rxjs';

@Component({
  selector: 'app-search',
  standalone: true,
  imports: [MatIconModule, NgIf, MatInputModule, MatFormFieldModule, ReactiveFormsModule],
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class SearchComponent implements OnChanges {
  searchControl = new FormControl('', { nonNullable: true });

  @Input() searchTerm!: string | null;
  @Output() searchTermChange: Observable<string> = this.searchControl.valueChanges;

  ngOnChanges() {
    this.searchControl.setValue(this.searchTerm || '', { emitEvent: false });
  }

  clear() {
    this.searchControl.setValue('');
  }
}
