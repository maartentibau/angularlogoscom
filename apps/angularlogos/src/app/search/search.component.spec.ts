import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { ReactiveFormsModule } from '@angular/forms';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';

import { SearchComponent } from './search.component';

describe('SearchComponent', () => {
  let component: SearchComponent;
  let fixture: ComponentFixture<SearchComponent>;

  beforeEach(async () => {
    TestBed.configureTestingModule({
      imports: [NoopAnimationsModule, MatFormFieldModule, MatInputModule, ReactiveFormsModule, SearchComponent]
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(SearchComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('ngOnChanges', () => {
    it('should set a new value', () => {
      // prepare
      component.searchTerm = 'search-for';
      jest.spyOn(component.searchControl, 'setValue');

      // act
      component.ngOnChanges();

      // check
      expect(component.searchControl.setValue).toHaveBeenCalledWith(component.searchTerm, { emitEvent: false });
    });
  });

  describe('clear', () => {
    it('should clear the searchTerm', () => {
      // prepare
      component.searchControl.setValue('some-value');
      jest.spyOn(component.searchControl, 'setValue');

      // act
      component.clear();

      // check
      expect(component.searchControl.setValue).toHaveBeenCalledWith('');
      expect(component.searchControl.value).toBe('');
    });

    it('should emit empty string', (done) => {
      // check
      component.searchControl.valueChanges.subscribe((value: string) => {
        expect(value).toBe('');
        done();
      });

      // act
      component.clear();
    });
  });
});
