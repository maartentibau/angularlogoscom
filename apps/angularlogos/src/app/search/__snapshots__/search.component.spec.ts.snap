// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent should render 1`] = `
<app-search
  searchControl={[Function FormControl2]}
  searchTermChange={[Function EventEmitter_]}
>
  <mat-form-field
    class="mat-mdc-form-field"
    subscriptsizing="dynamic"
  >
    <div
      class="mat-mdc-text-field-wrapper mdc-text-field"
    >
      <div
        class="mat-mdc-form-field-flex"
      >
        <div
          class="mat-mdc-form-field-infix"
        >
          <input
            class="mat-mdc-input-element"
            matinput=""
            subscriptsizing="0"
          />
        </div>
      </div>
    </div>
    <div
      class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
    >
    </div>
  </mat-form-field>
</app-search>
`;
